mtags/src/main/scala/scala/meta/internal/mtags/AtomicTrieMap.scala

```diff
@@ -0,0 +1,41 @@
+package scala.meta.internal.mtags
+
+import java.util.concurrent.ConcurrentHashMap
+
+import scala.collection.concurrent.TrieMap
+
+/**
+ * This class is a wrapper around TrieMap that provides atomic updateWith
+ */
+final class AtomicTrieMap[K, V] {
+  private val trieMap = new TrieMap[K, V]()
+  private val concurrentMap = new ConcurrentHashMap[K, V]
+
+  def get(key: K): Option[V] = trieMap.get(key)
+
+  def contains(key: K): Boolean = trieMap.contains(key)
+
+  def updateWith(key: K)(remappingFunc: Option[V] => Option[V]): Unit = {
+    val computeFunction = new java.util.function.BiFunction[K, V, V] {
+      override def apply(k: K, v: V): V = {
+        trieMap.get(key) match {
+          case Some(value) =>
+            remappingFunc(Some(value)) match {
+              case Some(newValue) =>
+                trieMap.update(key, newValue)
+              case None =>
+                trieMap.remove(key)
+            }
+          case None =>
+            remappingFunc(None).foreach(trieMap.update(key, _))
+        }
+        null.asInstanceOf[V]
+      }
+    }
+    concurrentMap.compute(key, computeFunction)
```

Comment:
Doesn't this duplicate the amount of data we hold? Should we just use ConcurrentMap instead?

Reply:
Size of concurrentMap is always 0, its only used to provide atomic updateWith for trieMap.
I don't think we should simply replace TrieMap with ConcurrentMap, because `get` on TrieMap should be faster

Reply:
> Size of concurrentMap is always 0, its only used to provide atomic updateWith for trieMap.

Can you ELI5 this? :D 

> get on TrieMap should be faster

Can you share some source about that, benchmark maybe? I did some reading, found https://medium.com/@igabaydulin/java-concurrenthashmap-vs-scala-concurrent-map-e185e8a0b798 which then led me to https://github.com/scala/scala/pull/10027 and I need to read everything once again because I'm already lost. Best part - those are not even about TrieMap 😅